### PR TITLE
Add support for JSON output writing

### PR DESCRIPTION
**Pull request by davidkroell**
Adding a new flag `-format` to set the output writer formatting. The default behavior keeps the same, as `text` is the default format.

Adding this functionality required to refactor the code a bit. Logic for filtering (top/over) and writing is now moved to the `gocognit` package.

Also updated to go1.18

### CODE REVIEW

File: cmd/gocognit/main.go [hunk 1]
```diff
@@ -119,15 +133,41 @@ func analyzeDir(dirname string, stats []gocognit.Stat) []gocognit.Stat {
...
+	}
+
+	filtered := filter.Apply(sortedStats)
+
+	err := formatter.Write(filtered)
+	if err != nil {
+		panic(err)
+	}
+
 	return len(sortedStats)
```
Comment (uudashr) [inline cmd/gocognit/main.go:171]: This should returned the filtered stats? Initially the return value used on `main()` to indicate wether we have stats written or not.

With your changes, the exit code not only decided  in `main()` function, but also in `writeStats(..)`. Do you have any idea to make it better?

**Review by uudashr — CHANGES_REQUESTED** [review]
Have you test the behaviors of the command line? We need to know whether the behavior is backward compatible or not. Since there might be some people utilized the command line for automation

**Comment by uudashr** [comment]
I've made some changes on master.
The Changes: `log.Fatal(..)` and `os.Exit(..)` only happen in `main()` function

Could you rebase from master, it might help to ensure the behavior of command line is backward compatible

File: cmd/gocognit/main.go [hunk 2]
```diff
@@ -38,19 +38,33 @@ Flags:
...
-	top  = flag.Int("top", -1, "show the top N most complex functions only")
-	avg  = flag.Bool("avg", false, "show the average complexity")
+	supportedFormats = []string{
+		textFormat, jsonFormat, jsonPrettyFormat,
+	}
+
+	over   = flag.Int("over", defaultValueIndicator, "show functions with complexity > N only")
+	top    = flag.Int("top", defaultValueIndicator, "show the top N most complex functions only")
+	avg    = flag.Bool("avg", false, "show the average complexity")
+	format = flag.String("format", "text", fmt.Sprintf("which format to use, supported formats: %v", supportedFormats))
```
[the remark this answers is not in the document]
Comment (davidkroell) [reply cmd/gocognit/main.go]: ofc, haven't seen that - thanks

File: cmd/gocognit/main.go (hunk 1 above)
Comment (davidkroell) [reply cmd/gocognit/main.go]: Yes, I've now refactored it a bit. Nevertheless this should be a very rare case.

**Comment by davidkroell** [comment]
I've refactored it and also did the rebase. Please review again

File: cmd/gocognit/main.go [hunk 3]
```diff
@@ -149,17 +166,43 @@ func analyzeDir(dirname string, stats []gocognit.Stat) ([]gocognit.Stat, error)
...
+		break
+	case jsonFormat:
+		formatter = gocognit.NewJsonFormatter(w, false)
+		break
+	case jsonPrettyFormat:
+		formatter = gocognit.NewJsonFormatter(w, true)
+		break
+	default:
+		fmt.Printf("Format '%s' is not valid, use a supported format %v", *format, supportedFormats)
+		os.Exit(1)
```
Comment (uudashr) [inline cmd/gocognit/main.go:195]: Let's put all the exit (os.Exit or log.Fatal) on the `func main()`. What if we return an error?

File: formatter.go [hunk 4]
```diff
@@ -0,0 +1,5 @@
+package gocognit
+
+type Formatter interface {
+	Write([]Stat) error
+}
```
Comment (uudashr) [inline formatter.go:5]: I'm thinking about the Golang idiom, do you think make sense if we follow them? see https://pkg.go.dev/io

Reader interface -> Read
Writer interface -> Write
Seeker interface -> Seek
Closer interface -> Close

**Review by uudashr — CHANGES_REQUESTED** [review]

**Comment by uudashr** [comment]
Your test is fail on 1.16. Do you have any specific reason why you upgrade the go version to 1.18?

**Comment by uudashr** [comment]
Change back the go to 1.16 should make it work

**Comment by uudashr** [comment]
@davidkroell May I know why you need this json format?

**Review by uudashr — APPROVED** [review]

**Comment by davidkroell** [comment]
I need this because I'd like to build a suite of quality checkers like this and generate reports based on these results and therefore json is an easy format to work with.
Thanks for merging!

**Comment by uudashr** [comment]
Hi @davidkroell, do you really need feature for for non-pretty JSON? Or any kind of JSON output is fine?

**Comment by davidkroell** [comment]
Hi @uudashr any JSON output is fine.

**Pull request merged**